File: actool/bom.py
"""
BOM (Bill of Materials) file writer.

BOM is Apple's container format used for .car files, .pkg installers, etc.
All BOM structures use big-endian byte order.
"""

import struct


class BOMWriter:
    """Writes BOM format files."""

    def __init__(self):
        self._blocks: list[bytes] = [b""]  # Block 0 is always empty
        self._named_blocks: dict[str, int] = {}
        self._trees: dict[str, int] = {}

    def add_block(self, data: bytes) -> int:
        """Add a block and return its index."""
        idx = len(self._blocks)
        self._blocks.append(data)
        return idx

    def add_named_block(self, name: str, data: bytes) -> int:
        """Add a named block (variable)."""
        idx = self.add_block(data)
        self._named_blocks[name] = idx
        return idx

    def add_tree(self, name: str, entries: list[tuple[bytes, bytes]],
                 block_size: int = 4096) -> int:
        """Add a BOM tree with the given key-value entries.

        entries: list of (key_bytes, value_bytes) pairs.
        Returns the block index of the tree header.
        """
        # Build leaf nodes. Each leaf node can hold up to block_size worth of entries.
        # Entry format in node: value_block_idx(4) + key_block_idx(4) = 8 bytes per entry
        # Node header: isLeaf(2) + count(2) + forward(4) + backward(4) = 12 bytes
        max_entries_per_node = (block_size - 12) // 8

        if len(entries) == 0:
            # Empty tree - single empty leaf node
            node_data = struct.pack(">HHiI", 1, 0, 0, 0)
            node_idx = self.add_block(node_data)
            tree_header = struct.pack(">4sIIIIBII", b"tree", 1, node_idx,
                                      block_size, 0, 0, 0, 0)
            tree_idx = self.add_block(tree_header)
            self._named_blocks[name] = tree_idx
            return tree_idx

        # Create leaf nodes
        leaf_nodes = []
        for start in range(0, len(entries), max_entries_per_node):
            batch = entries[start:start + max_entries_per_node]
            leaf_nodes.append(batch)

        if len(leaf_nodes) == 1:
            # Single leaf node - simple case
            node_data = self._build_leaf_node(leaf_nodes[0], 0, 0, block_size)
            node_idx = self.add_block(node_data)
            tree_header = struct.pack(">4sIIIIBII", b"tree", 1, node_idx,
                                      block_size, len(entries), 0, 0, 0)
            tree_idx = self.add_block(tree_header)
            self._named_blocks[name] = tree_idx
            return tree_idx

        # Multiple leaf nodes - need internal node(s)
        # Two-pass: reserve block indices first, then build with correct links
        leaf_indices = []
        for batch in leaf_nodes:
            # Reserve a block index with placeholder data
            leaf_indices.append(self.add_block(b"\x00"))

        # Now build each leaf with correct forward/backward links
        for i, batch in enumerate(leaf_nodes):
            fwd = leaf_indices[i + 1] if i + 1 < len(leaf_indices) else 0
            bwd = leaf_indices[i - 1] if i > 0 else 0
            node_data = self._build_leaf_node(batch, fwd, bwd, block_size)
            # Replace the placeholder block
            self._blocks[leaf_indices[i]] = node_data

        # Build internal node
        # Internal node format: isLeaf(2) + count(2) + forward(4) + backward(4)
        # Then: child0(4), [key_block(4) + child(4)] * count
        internal = struct.pack(">HHII", 0, len(leaf_indices) - 1, 0, 0)
        internal += struct.pack(">I", leaf_indices[0])
        for i in range(1, len(leaf_indices)):
            # Key is the first key of this leaf node
            first_key = leaf_nodes[i][0][0]
            key_idx = self.add_block(first_key)
            internal += struct.pack(">II", key_idx, leaf_indices[i])
        internal_idx = self.add_block(internal)

        tree_header = struct.pack(">4sIIIIBII", b"tree", 1, internal_idx,
                                  block_size, len(entries), 0, 0, 0)
        tree_idx = self.add_block(tree_header)
        self._named_blocks[name] = tree_idx
        return tree_idx

    def _build_leaf_node(self, entries: list[tuple[bytes, bytes]],
                         forward: int, backward: int,
                         block_size: int = 4096) -> bytes:
        """Build a leaf node from entries.

        Apple's BOM trees embed key data inline in the leaf node.  The layout
        is: header + entry pairs + padding to *block_size* + inline key data
        appended AFTER the block_size boundary.  The CoreUI tree reader reads
        block_size bytes for the entry section, then reads inline keys starting
        at offset block_size.  Getting this wrong causes value lookups to
        return garbage (e.g. garbled color components or multisize descriptors).
        """
        node = struct.pack(">HHII", 1, len(entries), forward, backward)
        key_data_list = []
        for key_data, value_data in entries:
            val_idx = self.add_block(value_data)
            key_idx = self.add_block(key_data)
            node += struct.pack(">II", val_idx, key_idx)
            key_data_list.append(key_data)
        # Pad entry section to exactly block_size
        if len(node) < block_size:
            node += b"\x00" * (block_size - len(node))
        # Append inline key data AFTER the block_size boundary
        for kd in key_data_list:
            node += kd
        return node

    def add_raw_key_tree(self, name: str,
                         entries: list[tuple[int, bytes]],
                         block_size: int = 1024) -> int:
        """Add a BOM tree where keys are raw uint32 values (not block refs).

        Used for BITMAPKEYS where the key is the facet identifier stored
        directly as a uint32 in the node entry.

        entries: list of (raw_key_uint32, value_bytes) pairs, sorted by key.
        """
        if not entries:
            return self.add_tree(name, [], block_size)

        # Build leaf node(s)
        # Entry format: value_block_idx(4) + raw_key(4) = 8 bytes
        max_per_node = (block_size - 12) // 8

        leaf_batches = []
        for start in range(0, len(entries), max_per_node):
            leaf_batches.append(entries[start:start + max_per_node])

        if len(leaf_batches) == 1:
            node = struct.pack(">HHII", 1, len(entries), 0, 0)
            for raw_key, value_data in entries:
                val_idx = self.add_block(value_data)
                node += struct.pack(">II", val_idx, raw_key)
            # Pad node to block_size (BOM reader may read full block_size)
            if len(node) < block_size:
                node += b"\x00" * (block_size - len(node))
            node_idx = self.add_block(node)
            tree_header = struct.pack(">4sIIIIBII", b"tree", 1, node_idx,
                                      block_size, len(entries), 1, 0, 0)
            tree_idx = self.add_block(tree_header)
            self._named_blocks[name] = tree_idx
            return tree_idx

        # Multiple leaf nodes — same two-pass approach
        leaf_indices = []
        for batch in leaf_batches:
            leaf_indices.append(self.add_block(b"\x00"))

        for i, batch in enumerate(leaf_batches):
            fwd = leaf_indices[i + 1] if i + 1 < len(leaf_indices) else 0
            bwd = leaf_indices[i - 1] if i > 0 else 0
            node = struct.pack(">HHII", 1, len(batch), fwd, bwd)
            for raw_key, value_data in batch:
                val_idx = self.add_block(value_data)
                node += struct.pack(">II", val_idx, raw_key)
            if len(node) < block_size:
                node += b"\x00" * (block_size - len(node))
            self._blocks[leaf_indices[i]] = node

        internal = struct.pack(">HHII", 0, len(leaf_indices) - 1, 0, 0)
        internal += struct.pack(">I", leaf_indices[0])
        for i in range(1, len(leaf_indices)):
            first_key = leaf_batches[i][0][0]
            internal += struct.pack(">II", first_key, leaf_indices[i])
        internal_idx = self.add_block(internal)

        tree_header = struct.pack(">4sIIIIBII", b"tree", 1, internal_idx,
                                  block_size, len(entries), 1, 0, 0)
        tree_idx = self.add_block(tree_header)
        self._named_blocks[name] = tree_idx
        return tree_idx

    def write(self, path: str):
        """Write the BOM file to disk."""
        MIN_TABLE_SIZE = 256  # Apple pre-allocates at least 256 entries
        FREELIST_SIZE = 20    # 20-byte freelist trailer (all zeros)

        # Calculate block offsets
        header_size = 32  # BOMStore header
        current_offset = header_size
        num_used_blocks = len(self._blocks)  # Max block index + 1 (includes null block 0)

        block_entries = []
        for i, block_data in enumerate(self._blocks):
            if i == 0:
                block_entries.append((0, 0))
                continue
            if current_offset % 16 != 0:
                current_offset += 16 - (current_offset % 16)
            block_entries.append((current_offset, len(block_data)))
            current_offset += len(block_data)

        # Pad block table to minimum size with null entries
        while len(block_entries) < MIN_TABLE_SIZE:
            block_entries.append((0, 0))

        # Build index (block table + freelist)
        index_data = struct.pack(">I", len(block_entries))
        for offset, length in block_entries:
            index_data += struct.pack(">II", offset, length)
        index_data += b"\x00" * FREELIST_SIZE  # Empty freelist

        # Build vars section
        vars_data = struct.pack(">I", len(self._named_blocks))
        for name, block_idx in self._named_blocks.items():
            name_bytes = name.encode("ascii")
            vars_data += struct.pack(">IB", block_idx, len(name_bytes))
            vars_data += name_bytes

        # Align current_offset for vars
        if current_offset % 16 != 0:
            current_offset += 16 - (current_offset % 16)
        vars_offset = current_offset
        vars_length = len(vars_data)
        current_offset += vars_length

        # Align for index
        if current_offset % 16 != 0:
            current_offset += 16 - (current_offset % 16)
        index_offset = current_offset
        index_length = len(index_data)

        # Write the file
        with open(path, "wb") as f:
            # Header
            f.write(b"BOMStore")
            f.write(struct.pack(">I", 1))  # version
            f.write(struct.pack(">I", num_used_blocks))  # numberOfBlocks (used only)
            f.write(struct.pack(">I", index_offset))
            f.write(struct.pack(">I", index_length))
            f.write(struct.pack(">I", vars_offset))
            f.write(struct.pack(">I", vars_length))

            # Blocks
            for i, block_data in enumerate(self._blocks):
                if i == 0:
                    continue
                expected_offset = block_entries[i][0]
                current_pos = f.tell()
                if current_pos < expected_offset:
                    f.write(b"\x00" * (expected_offset - current_pos))
                f.write(block_data)

            # Vars
            current_pos = f.tell()
            if current_pos < vars_offset:
                f.write(b"\x00" * (vars_offset - current_pos))
            f.write(vars_data)

            # Index
            current_pos = f.tell()
            if current_pos < index_offset:
                f.write(b"\x00" * (index_offset - current_pos))
            f.write(index_data)

File: src/bom.rs
//! BOM (Bill of Materials) file writer.
//!
//! BOM is Apple's container format used for `.car` files and `.pkg`
//! installers. All structures are big-endian.

use byteorder::{BigEndian, WriteBytesExt};
use indexmap::IndexMap;
use std::io::{Seek, SeekFrom, Write};
use std::path::Path;

pub struct BomWriter {
    blocks: Vec<Vec<u8>>,
    named_blocks: IndexMap<String, usize>,
    inline_key_size: Option<usize>,
}

impl BomWriter {
    pub fn new() -> Self {
        Self {
            blocks: vec![Vec::new()], // block 0 is always empty
            named_blocks: IndexMap::new(),
            inline_key_size: None,
        }
    }

    /// Set the fixed key size for subsequent `add_tree` calls. When set, the
    /// leaf node reserves `n_entries * key_size` zero bytes after the
    /// block_size padding, matching Apple's RENDITIONS leaf layout. Set back
    /// to `None` before adding variable-key trees (FACETKEYS, APPEARANCEKEYS).
    pub fn set_inline_key_size(&mut self, size: Option<usize>) {
        self.inline_key_size = size;
    }

    pub fn add_block(&mut self, data: Vec<u8>) -> usize {
        let idx = self.blocks.len();
        self.blocks.push(data);
        idx
    }

    pub fn add_named_block(&mut self, name: &str, data: Vec<u8>) -> usize {
        let idx = self.add_block(data);
        self.named_blocks.insert(name.to_string(), idx);
        idx
    }

    pub fn add_tree(
        &mut self,
        name: &str,
        entries: &[(Vec<u8>, Vec<u8>)],
        block_size: u32,
    ) -> usize {
        // Entry format: value_block_idx(4) + key_block_idx(4) = 8 bytes
        // Node header: isLeaf(2) + count(2) + forward(4) + backward(4) = 12 bytes
        let max_per_node = ((block_size as usize) - 12) / 8;

        // Fixed-key trees (RENDITIONS) use the caller-supplied key width.
        // Variable-key trees (FACETKEYS, APPEARANCEKEYS) still carry an
        // inline-key region whose width equals the longest key in the tree;
        // CUICatalog reads that region when enumerating facets/appearances.
        let effective_key_size = self.inline_key_size.unwrap_or_else(|| {
            entries.iter().map(|(k, _)| k.len()).max().unwrap_or(0)
        });

        if entries.is_empty() {
            let mut node = Vec::with_capacity(block_size as usize);
            node.write_u16::<BigEndian>(1).unwrap();
            node.write_u16::<BigEndian>(0).unwrap();
            node.write_i32::<BigEndian>(0).unwrap();
            node.write_u32::<BigEndian>(0).unwrap();
            node.resize(block_size as usize, 0);
            let node_idx = self.add_block(node);
            let tree_hdr =
                build_tree_header(node_idx, block_size, 0, 0, effective_key_size as u32);
            let tree_idx = self.add_block(tree_hdr);
            self.named_blocks.insert(name.to_string(), tree_idx);
            return tree_idx;
        }

        let leaf_batches: Vec<&[(Vec<u8>, Vec<u8>)]> =
            entries.chunks(max_per_node).collect();

        if leaf_batches.len() == 1 {
            let node = self.build_leaf_node(
                leaf_batches[0],
                0,
                0,
                block_size,
                Some(effective_key_size),
            );
            let node_idx = self.add_block(node);
            let tree_hdr = build_tree_header(
                node_idx,
                block_size,
                entries.len() as u32,
                0,
                effective_key_size as u32,
            );
            let tree_idx = self.add_block(tree_hdr);
            self.named_blocks.insert(name.to_string(), tree_idx);
            return tree_idx;
        }

        // Reserve leaf indices before we know their contents.
        let leaf_indices: Vec<usize> = leaf_batches
            .iter()
            .map(|_| self.add_block(vec![0]))
            .collect();

        for (i, batch) in leaf_batches.iter().enumerate() {
            let fwd = if i + 1 < leaf_indices.len() {
                leaf_indices[i + 1] as u32
            } else {
                0
            };
            let bwd = if i > 0 { leaf_indices[i - 1] as u32 } else { 0 };
            let node = self.build_leaf_node(batch, fwd, bwd, block_size, Some(effective_key_size));
            self.blocks[leaf_indices[i]] = node;
        }

        let mut internal = Vec::new();
        internal.write_u16::<BigEndian>(0).unwrap(); // isLeaf = 0
        internal
            .write_u16::<BigEndian>((leaf_indices.len() - 1) as u16)
            .unwrap();
        internal.write_u32::<BigEndian>(0).unwrap();
        internal.write_u32::<BigEndian>(0).unwrap();
        internal
            .write_u32::<BigEndian>(leaf_indices[0] as u32)
            .unwrap();
        for i in 1..leaf_indices.len() {
            let first_key = leaf_batches[i][0].0.clone();
            let key_idx = self.add_block(first_key);
            internal.write_u32::<BigEndian>(key_idx as u32).unwrap();
            internal
                .write_u32::<BigEndian>(leaf_indices[i] as u32)
                .unwrap();
        }
        let internal_idx = self.add_block(internal);
        let tree_hdr = build_tree_header(
            internal_idx,
            block_size,
            entries.len() as u32,
            0,
            effective_key_size as u32,
        );
        let tree_idx = self.add_block(tree_hdr);
        self.named_blocks.insert(name.to_string(), tree_idx);
        tree_idx
    }


    fn build_leaf_node(
        &mut self,
        entries: &[(Vec<u8>, Vec<u8>)],
        forward: u32,
        backward: u32,
        block_size: u32,
        inline_key_size: Option<usize>,
    ) -> Vec<u8> {
        let mut node = Vec::new();
        node.write_u16::<BigEndian>(1).unwrap(); // isLeaf
        node.write_u16::<BigEndian>(entries.len() as u16).unwrap();
        node.write_u32::<BigEndian>(forward).unwrap();
        node.write_u32::<BigEndian>(backward).unwrap();
        let mut key_blob: Vec<u8> = Vec::new();
        for (key_data, value_data) in entries {
            let val_idx = self.add_block(value_data.clone());
            let key_idx = self.add_block(key_data.clone());
            node.write_u32::<BigEndian>(val_idx as u32).unwrap();
            node.write_u32::<BigEndian>(key_idx as u32).unwrap();
            if let Some(ks) = inline_key_size {
                key_blob.extend_from_slice(key_data);
                let pad = ks.saturating_sub(key_data.len());
                key_blob.resize(key_blob.len() + pad, 0);
            }
        }
        // For fixed-key trees (RENDITIONS) and variable-key trees whose keys
        // are padded to a common width, Apple inlines each key immediately
        // after the entries. A 4-byte zero gap separates the entry table from
        // the inline-key region.
        if !key_blob.is_empty() {
            node.extend_from_slice(&[0u8; 4]);
            node.extend_from_slice(&key_blob);
        }
        if (node.len() as u32) < block_size {
            node.resize(block_size as usize, 0);
        }
        // After the inline-key region the leaf is padded with zeros until
        // total length = block_size + n_entries * fixed_key_size.
        if let Some(ks) = inline_key_size {
            let target = block_size as usize + entries.len() * ks;
            if node.len() < target {
                node.resize(target, 0);
            }
        }
        node
    }

    pub fn add_raw_key_tree(
        &mut self,
        name: &str,
        entries: &[(u32, Vec<u8>)],
        block_size: u32,
    ) -> usize {
        if entries.is_empty() {
            let mut node = Vec::with_capacity(block_size as usize);
            node.write_u16::<BigEndian>(1).unwrap();
            node.write_u16::<BigEndian>(0).unwrap();
            node.write_u32::<BigEndian>(0).unwrap();
            node.write_u32::<BigEndian>(0).unwrap();
            node.resize(block_size as usize, 0);
            let node_idx = self.add_block(node);
            let tree_hdr = build_tree_header(node_idx, block_size, 0, 1, 0);
            let tree_idx = self.add_block(tree_hdr);
            self.named_blocks.insert(name.to_string(), tree_idx);
            return tree_idx;
        }

        let max_per_node = ((block_size as usize) - 12) / 8;
        let batches: Vec<&[(u32, Vec<u8>)]> =
            entries.chunks(max_per_node).collect();

        if batches.len() == 1 {
            let mut node = Vec::new();
            node.write_u16::<BigEndian>(1).unwrap();
            node.write_u16::<BigEndian>(entries.len() as u16).unwrap();
            node.write_u32::<BigEndian>(0).unwrap();
            node.write_u32::<BigEndian>(0).unwrap();
            for (raw_key, value_data) in entries {
                let val_idx = self.add_block(value_data.clone());
                node.write_u32::<BigEndian>(val_idx as u32).unwrap();
                node.write_u32::<BigEndian>(*raw_key).unwrap();
            }
            if (node.len() as u32) < block_size {
                node.resize(block_size as usize, 0);
            }
            let node_idx = self.add_block(node);
            let tree_hdr =
                build_tree_header(node_idx, block_size, entries.len() as u32, 1, 0);
            let tree_idx = self.add_block(tree_hdr);
            self.named_blocks.insert(name.to_string(), tree_idx);
            return tree_idx;
        }

        let leaf_indices: Vec<usize> =
            batches.iter().map(|_| self.add_block(vec![0])).collect();

        for (i, batch) in batches.iter().enumerate() {
            let fwd = if i + 1 < leaf_indices.len() {
                leaf_indices[i + 1] as u32
            } else {
                0
            };
            let bwd = if i > 0 { leaf_indices[i - 1] as u32 } else { 0 };
            let mut node = Vec::new();
            node.write_u16::<BigEndian>(1).unwrap();
            node.write_u16::<BigEndian>(batch.len() as u16).unwrap();
            node.write_u32::<BigEndian>(fwd).unwrap();
            node.write_u32::<BigEndian>(bwd).unwrap();
            for (raw_key, value_data) in *batch {
                let val_idx = self.add_block(value_data.clone());
                node.write_u32::<BigEndian>(val_idx as u32).unwrap();
                node.write_u32::<BigEndian>(*raw_key).unwrap();
            }
            if (node.len() as u32) < block_size {
                node.resize(block_size as usize, 0);
            }
            self.blocks[leaf_indices[i]] = node;
        }

        let mut internal = Vec::new();
        internal.write_u16::<BigEndian>(0).unwrap();
        internal
            .write_u16::<BigEndian>((leaf_indices.len() - 1) as u16)
            .unwrap();
        internal.write_u32::<BigEndian>(0).unwrap();
        internal.write_u32::<BigEndian>(0).unwrap();
        internal
            .write_u32::<BigEndian>(leaf_indices[0] as u32)
            .unwrap();
        for i in 1..leaf_indices.len() {
            let first_key = batches[i][0].0;
            internal.write_u32::<BigEndian>(first_key).unwrap();
            internal
                .write_u32::<BigEndian>(leaf_indices[i] as u32)
                .unwrap();
        }
        let internal_idx = self.add_block(internal);
        let tree_hdr =
            build_tree_header(internal_idx, block_size, entries.len() as u32, 1, 0);
        let tree_idx = self.add_block(tree_hdr);
        self.named_blocks.insert(name.to_string(), tree_idx);
        tree_idx
    }

    pub fn write<P: AsRef<Path>>(&self, path: P) -> std::io::Result<()> {
        let buf = self.to_bytes();
        std::fs::write(path, buf)
    }

    pub fn to_bytes(&self) -> Vec<u8> {
        const MIN_TABLE_SIZE: usize = 256;
        const FREELIST_SIZE: usize = 20;

        let header_size: u64 = 32;
        let mut current_offset: u64 = header_size;
        let num_used_blocks = self.blocks.len() as u32;

        let mut block_entries: Vec<(u32, u32)> = Vec::new();
        for (i, block_data) in self.blocks.iter().enumerate() {
            if i == 0 {
                block_entries.push((0, 0));
                continue;
            }
            if current_offset % 16 != 0 {
                current_offset += 16 - (current_offset % 16);
            }
            block_entries
                .push((current_offset as u32, block_data.len() as u32));
            current_offset += block_data.len() as u64;
        }
        while block_entries.len() < MIN_TABLE_SIZE {
            block_entries.push((0, 0));
        }

        let mut index_data = Vec::new();
        index_data
            .write_u32::<BigEndian>(block_entries.len() as u32)
            .unwrap();
        for (off, len) in &block_entries {
            index_data.write_u32::<BigEndian>(*off).unwrap();
            index_data.write_u32::<BigEndian>(*len).unwrap();
        }
        index_data.extend(std::iter::repeat(0u8).take(FREELIST_SIZE));

        let mut vars_data = Vec::new();
        vars_data
            .write_u32::<BigEndian>(self.named_blocks.len() as u32)
            .unwrap();
        for (name, block_idx) in &self.named_blocks {
            let name_bytes = name.as_bytes();
            vars_data.write_u32::<BigEndian>(*block_idx as u32).unwrap();
            vars_data.write_u8(name_bytes.len() as u8).unwrap();
            vars_data.extend_from_slice(name_bytes);
        }

        if current_offset % 16 != 0 {
            current_offset += 16 - (current_offset % 16);
        }
        let vars_offset = current_offset;
        let vars_length = vars_data.len() as u32;
        current_offset += vars_length as u64;

        if current_offset % 16 != 0 {
            current_offset += 16 - (current_offset % 16);
        }
        let index_offset = current_offset;
        let index_length = index_data.len() as u32;

        let mut out = std::io::Cursor::new(Vec::new());
        out.write_all(b"BOMStore").unwrap();
        out.write_u32::<BigEndian>(1).unwrap(); // version
        out.write_u32::<BigEndian>(num_used_blocks).unwrap();
        out.write_u32::<BigEndian>(index_offset as u32).unwrap();
        out.write_u32::<BigEndian>(index_length).unwrap();
        out.write_u32::<BigEndian>(vars_offset as u32).unwrap();
        out.write_u32::<BigEndian>(vars_length).unwrap();

        for (i, block_data) in self.blocks.iter().enumerate() {
            if i == 0 {
                continue;
            }
            let expected = block_entries[i].0 as u64;
            let pos = out.position();
            if pos < expected {
                let pad = (expected - pos) as usize;
                out.write_all(&vec![0u8; pad]).unwrap();
            }
            out.write_all(block_data).unwrap();
        }

        let pos = out.position();
        if pos < vars_offset {
            let pad = (vars_offset - pos) as usize;
            out.write_all(&vec![0u8; pad]).unwrap();
        }
        out.write_all(&vars_data).unwrap();

        let pos = out.position();
        if pos < index_offset {
            let pad = (index_offset - pos) as usize;
            out.write_all(&vec![0u8; pad]).unwrap();
        }
        out.write_all(&index_data).unwrap();

        let _ = out.seek(SeekFrom::Start(0));
        out.into_inner()
    }
}

impl Default for BomWriter {
    fn default() -> Self {
        Self::new()
    }
}

fn build_tree_header(
    root_idx: usize,
    block_size: u32,
    n_paths: u32,
    flag: u8,
    key_size: u32,
) -> Vec<u8> {
    // "tree"(4) + version(4) + root(4) + block_size(4) + n_paths(4) +
    // flag(1) + key_size(4) + pad(4).
    // key_size is the per-entry inline key length: the actual size for
    // fixed-key trees (RENDITIONS), 0xFFFFFFFF for variable-key trees
    // (FACETKEYS, APPEARANCEKEYS), or 0 for raw-key trees (BITMAPKEYS).
    let mut hdr = Vec::new();
    hdr.extend_from_slice(b"tree");
    hdr.write_u32::<BigEndian>(1).unwrap();
    hdr.write_u32::<BigEndian>(root_idx as u32).unwrap();
    hdr.write_u32::<BigEndian>(block_size).unwrap();
    hdr.write_u32::<BigEndian>(n_paths).unwrap();
    hdr.write_u8(flag).unwrap();
    hdr.write_u32::<BigEndian>(key_size).unwrap();
    hdr.write_u32::<BigEndian>(0).unwrap();
    hdr
}

#[cfg(test)]
mod tests {
    use super::*;
    use byteorder::ReadBytesExt;
    use std::io::Cursor;

    #[test]
    fn empty_bom() {
        let bom = BomWriter::new();
        let data = bom.to_bytes();
        assert_eq!(&data[..8], b"BOMStore");
        let mut c = Cursor::new(&data[8..]);
        assert_eq!(c.read_u32::<BigEndian>().unwrap(), 1); // version
        assert_eq!(c.read_u32::<BigEndian>().unwrap(), 1); // num blocks (just null)
    }

    #[test]
    fn single_named_block() {
        let mut bom = BomWriter::new();
        let idx = bom.add_named_block("HELLO", b"world data".to_vec());
        assert_eq!(idx, 1);
        let data = bom.to_bytes();
        assert_eq!(&data[..8], b"BOMStore");
        // The bytes "world data" must appear somewhere in the output
        assert!(data.windows(10).any(|w| w == b"world data"));
    }

    #[test]
    fn tree_single_leaf() {
        let mut bom = BomWriter::new();
        let entries = vec![(b"key1".to_vec(), b"val1".to_vec())];
        bom.add_tree("TREE", &entries, 4096);
        let data = bom.to_bytes();
        assert!(data.windows(4).any(|w| w == b"tree"));
        assert!(data.windows(4).any(|w| w == b"key1"));
        assert!(data.windows(4).any(|w| w == b"val1"));
    }

    #[test]
    fn raw_key_tree() {
        let mut bom = BomWriter::new();
        let entries: Vec<(u32, Vec<u8>)> =
            vec![(1, b"aaaa".to_vec()), (2, b"bbbb".to_vec())];
        bom.add_raw_key_tree("TBL", &entries, 1024);
        let data = bom.to_bytes();
        assert!(data.windows(4).any(|w| w == b"aaaa"));
        assert!(data.windows(4).any(|w| w == b"bbbb"));
    }

    #[test]
    fn named_block_ordering_preserved() {
        let mut bom = BomWriter::new();
        bom.add_named_block("FIRST", b"AAAA".to_vec());
        bom.add_named_block("SECOND", b"BBBB".to_vec());
        bom.add_named_block("THIRD", b"CCCC".to_vec());
        let data = bom.to_bytes();
        // Find the vars section by looking for the name bytes.
        let first = data
            .windows(5)
            .position(|w| w == b"FIRST")
            .expect("FIRST not found");
        let second = data
            .windows(6)
            .position(|w| w == b"SECOND")
            .expect("SECOND not found");
        let third = data
            .windows(5)
            .position(|w| w == b"THIRD")
            .expect("THIRD not found");
        assert!(first < second);
        assert!(second < third);
    }
}
